src/rust/engine/process_execution/src/remote_cache.rs

```diff
@@ -450,47 +450,70 @@ impl crate::CommandRunner for CommandRunner {
     )
     .await?;
 
-    // Check the remote Action Cache to see if this request was already computed.
-    // If so, return immediately with the result.
-    if self.cache_read {
-      let response = with_workunit(
-        context.workunit_store.clone(),
-        "check_action_cache".to_owned(),
-        WorkunitMetadata::with_level(Level::Debug),
-        crate::remote::check_action_cache(
-          action_digest,
-          &self.metadata,
-          self.platform,
-          &context,
-          self.action_cache_client.clone(),
-          self.store.clone(),
-          self.eager_fetch,
-        ),
-        |_, md| md,
-      )
-      .await;
-      match response {
-        Ok(cached_response_opt) => {
-          log::debug!(
-            "remote cache response: digest={:?}: {:?}",
+    let mut local_execution_future = self.underlying.run(req, context.clone());
+
+    let result = if self.cache_read {
+      // A future to read from the cache and log the results accordingly.
+      let cache_read_future = async {
+        let response = with_workunit(
+          context.workunit_store.clone(),
+          "check_action_cache".to_owned(),
+          WorkunitMetadata::with_level(Level::Debug),
+          crate::remote::check_action_cache(
             action_digest,
+            &self.metadata,
+            self.platform,
+            &context,
+            self.action_cache_client.clone(),
+            self.store.clone(),
+            self.eager_fetch,
+          ),
+          |_, md| md,
+        )
+        .await;
+        match response {
+          Ok(cached_response_opt) => {
+            log::debug!(
+              "remote cache response: digest={:?}: {:?}",
+              action_digest,
+              cached_response_opt
+            );
+            if cached_response_opt.is_some() {
+              context
+                .workunit_store
+                .increment_counter(Metric::SpeculationRemoteCacheCompletedFirst, 1);
+            }
             cached_response_opt
-          );
+          }
+          Err(err) => {
+            log::warn!("Failed to read from remote cache: {}", err);
+            None
+          }
+        }
+      };
 
-          if let Some(cached_response) = cached_response_opt {
+      // We speculate between reading from the remote cache vs. running locally. If there was a
+      // cache hit, we return early because there will be no need to write to the cache. Otherwise,
+      // we run the process locally and will possibly write it to the cache later.
+      tokio::select! {
+        cache_result = cache_read_future => {
+          if let Some(cached_response) = cache_result {
             return Ok(cached_response);
+          } else {
+           local_execution_future.await?
```

Comment:
Oh, hm, @stuhood I think this might be bad behavior. If the remote cache was empty, and we found that out quickly, then we will cancel the local_execution select! arm and then restart the exact same process again. That sounds wrong to launch the process twice, right?

Reply:
This is "the same" process: the first time you awaited it, you did so mutably via the macro: the second time, you're just finishing waiting for it.

Reply:
Ahh, that makes sense. It's the same `Future` in both places. Thanks!